libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/NavigationStateCoordinator.kt

```diff
@@ -0,0 +1,48 @@
+package com.mapbox.navigation.dropin.coordinator
+
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.dropin.DropInNavigationViewContext
+import com.mapbox.navigation.dropin.component.navigationstate.NavigationState
+import com.mapbox.navigation.dropin.extensions.flowRoutesUpdated
+import com.mapbox.navigation.dropin.lifecycle.UIComponent
+import kotlinx.coroutines.flow.combine
+
+/**
+ * Coordinator that updates NavigationState as following:
+ *
+ * ```
+ *  [ FreeDrive ]━━( destination != null )━━━━>>[ RoutePreview ]<<━━━━━━━━━━━━┓
+ *       ^         ( & routes not empty  )             ┃                      ┃
+ *       ┃                                (activeNavigationStarted=true)      ┃
+ *       ┃                                             ┃                      ┃
+ *       ┃                                             ┃         (activeNavigationStarted=false)
+ *       ┣━━━━( destination == null )━━━━┓             v                      ┃
+ *       ┗━━━━( routes empty        )━━━━┻━━━[ ActiveNavigation ] ━━━━━━━━━━━━┛
+ * ```
+ */
+internal class NavigationStateCoordinator(
+    context: DropInNavigationViewContext
+) : UIComponent() {
```

Comment:
I'm only commenting on the naming here. `UICoordinator` is a class, and the definition of the "coordinator" is tied closer to that class than other possible philosophies we try to fit it to. 

We are calling other components a "behavior", that extend `UIComponent`. We should call this `NavigationStateBehavior`.

Reply:
To help me better understand differences between `Coordinator`, `Component` and `Behaviour` classes, can you give me a brief description for each type?

To be honest with you, "behaviour" also doesn't fit this class purpose. Behaviour classes are _Adjustment_ type of classes that when assigned to a target, changes or augments its basic functionality. This class observes 3 sources of truth (destination, routes and activeNavigationStarted) and moves NavigationState to a correct state by following above FSM. I think either "Controller" or "Manager" would be more appropriate name here.

Reply:
Linking some definitions https://github.com/mapbox/mapbox-navigation-android/pull/5521 in case someone comes in for history